src/shared/iterate-emitter.test.js

```diff
@@ -0,0 +1,141 @@
+import { EventEmitter } from 'events';
+
+// import { describe, it, expect } from '@jest/globals';
+
+import test from 'ava';
+
+import { iterateEmitter } from './iterate-emitter.js';
+
+const timeout = timeoutTestMiddleware();
+test.beforeEach(async () => {
+  await timeout.setup();
+});
+test.afterEach(async () => {
+  await timeout.teardown();
+});
+
+test('iterate values emitted by an EventEmitter', async t => {
+  const emitter = new EventEmitter();
+  timeout.thread(async ({ timeout }) => {
+    for (let i = 0; i < 100; i++) {
+      emitter.emit('next', i);
+      await timeout();
+    }
+  });
+  t.deepEqual(await take(10, iterateEmitter(emitter, 'next')), [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]);
+});
+
+test('stops iterating values emitted by an EventEmitter when it emits a completion event', async t => {
+  const emitter = new EventEmitter();
+  timeout.thread(async ({ timeout }) => {
+    for (let i = 0; i < 5; i++) {
+      emitter.emit('next', i);
+      await timeout();
+    }
+    emitter.emit('complete');
+    for (let i = 5; i < 100; i++) {
+      emitter.emit('next', i);
+      await timeout();
+    }
+  });
+  t.deepEqual(await take(10, iterateEmitter(emitter, 'next', 'complete')), [0, 1, 2, 3, 4]);
+});
+
+test('throws an error when an EventEmitter emits an error event', async t => {
+  const emitter = new EventEmitter();
+  timeout.thread(async ({ timeout }) => {
+    for (let i = 0; i < 5; i++) {
+      emitter.emit('next', i);
+      await timeout();
+    }
+    emitter.emit('error', new Error('error'));
+    for (let i = 0; i < 5; i++) {
+      emitter.emit('next', i);
+      await timeout();
+    }
+  });
+  await t.throwsAsync(
+    async () => await take(10, iterateEmitter(emitter, 'next', 'complete', 'error'))
+  );
+});
+
+test('iterate queued values emitted by an EventEmitter after it emits a completion event', async t => {
```

Comment:
`iterate queued ...` is concerned with the values before `emit('complete')` and `stops iterating ...` is concerned with the values after `emit('complete')`. `stop iterating ...` has a timeout after each emitted value such that when it `complete`s iterateEmitter's value queue will have been consumed. `iterate queued ...` without a timeout will `complete` when it's queue has 5 values.

Additionally `stop iterating ...` does not have a timeout after `complete`. That's important so that `emit('next', 5)` occurs. That will run iterateEmitter's next listener which should not queue 5 since `complete` has happened.

It is true that `emit('next', 6)` shouldn't happen. Maybe that second loop should be changed. But I'm not sure.